### PR TITLE
spring security

### PR DESCRIPTION
**Pull request by jcgrethe**
# Summary

- Spring security Implemented. Only creation of book and users are allowed without login
## Screenshot

Screenshot from 2019-07-31 18-05-48
Screenshot from 2019-07-31 18-12-56

### CODE REVIEW

File: src/main/java/wolox/training/controllers/UserController.java [hunk 1]
```diff
@@ -44,7 +51,9 @@ public User get(@PathVariable Long id) throws UserNotFoundException {
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
     public User create(@RequestBody User user) {
-        return userRepository.save(user);
+        User u = userRepository.save(user);
+        u.setPassword(passwordEncoder.encode(user.getPassword()));
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/controllers/UserController.java]: The encoder should be in the setter method, and the `.save()` method should be the line after setting the password :) you can return the value directly like this: 

`return userRepository.save(u)`

File: src/main/java/wolox/training/controllers/UserController.java [hunk 2]
```diff
@@ -44,7 +51,9 @@ public User get(@PathVariable Long id) throws UserNotFoundException {
...
+        return u;
     }
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/controllers/UserController.java]: The controller should include a separate method to update the password (other than the standard `update` method). It should receive the params `oldPassword` and `newPassword`, compare the current password and the old one and only update the password with the new value if they match. 

And the standard `update` method should update every field individually, except the password, keeping the current one.

File: src/main/java/wolox/training/models/User.java [hunk 3]
```diff
@@ -91,4 +103,13 @@ public void deleteBook(Book book) throws BookNonOwnedException {
     public Long getId() {
         return id;
     }
+
+    @Transient
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/models/User.java:110]: Include the encryption here 🙂 You could also add a password validation with Preconditions: not null, minimum amount of characters, whatever you prefer :)

File: src/main/java/wolox/training/config/SecurityConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            "/403");
+    }
+
+    @Bean
+    public PasswordEncoder passwordEncoder(){
+        PasswordEncoder encoder = new BCryptPasswordEncoder();
+        return encoder;
+    }
+
+    @Autowired
```
Comment (Jimenarosello) [inline src/main/java/wolox/training/config/SecurityConfiguration.java]: I think you shouldn't need this method, let me know if there is any specific reason you configured it :) Same for UserDetailService (which I think is only used here)

**Review by Jimenarosello — CHANGES_REQUESTED** [review]
Nice job 😄 a few comments :)

**Review by Jimenarosello — APPROVED** [review]

**Review by Gianni-Nicolas — APPROVED** [review]

**Pull request merged**